File: day02.py
from collections import Counter


with open('day02-input.txt', 'r') as input_file:
    ids = [id.strip() for id in input_file]

# For each ID count the number of times each character is found.
repeats_counters = (Counter(id).values() for id in ids)

double_occurrences = 0
triple_occurrences = 0
for repeats_counter in repeats_counters:
    if 2 in repeats_counter:
        double_occurrences += 1
    if 3 in repeats_counter:
        triple_occurrences += 1

# pt 1
print(double_occurrences * triple_occurrences)

for index, id in enumerate(ids):
    # Only compare against IDs after the current.
    ids_to_compare = ids[index + 1:]
    for other_id in ids_to_compare:
        # Find characters that are different, in the same position.
        diffs = [i for i in range(len(id)) if id[i] != other_id[i]]
        if len(diffs) == 1:
            diff_index = diffs[0]
            # pt 2
            # Rebuild string without the different character
            print(id[:diff_index] + id[diff_index + 1:])
            break

File: day02.js
const fs = require('fs')

const ids = fs.readFileSync('day02-input.txt', 'utf8')
  .trim()
  .split('\n')

// Array of obj {char: number of times found}
const occurrencesCounters = ids.map(id => {
  const asArray = id.split('')
  const charCount = asArray.reduce((counter, char) => {
    counter[char] = (counter[char] || 0) + 1
    return counter
  }, {})
  return charCount
})

const occurrencesValues = occurrencesCounters.map(Object.values)

const [doubleCount, tripleCount] = occurrencesValues.reduce(([double, triple], occurrences) => {
    if (occurrences.includes(2)) double += 1
    if (occurrences.includes(3)) triple += 1
    return [double, triple]
  }, [0, 0])
// pt 1
console.log(doubleCount * tripleCount)


const idsAndToCompare = ids.map((id, index) => {
  // Find what ids to compare this one against
  const idsToCompare = ids.slice(index + 1)
  return [id, idsToCompare]
})

for (let [id, idsToCompare] of idsAndToCompare) {
  for (otherId of idsToCompare) {
    // Get char diffs against current id
    const diff = otherId.split('').reduce((acc, char, i) => {
      if (char !== id[i]) acc.push(i)
      return acc
    }, [])

    if (diff.length === 1) {
      const diffIndex = diff[0]
      // pt 2
      // Rebuild string excluding the diff char
      console.log(id.slice(0, diffIndex) + id.slice(diffIndex + 1))
      break
    }
  }
}
